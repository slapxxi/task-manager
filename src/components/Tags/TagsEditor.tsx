import { Keys, Regex } from '@lib';
import { ID, Tag as ITag, UserCreatedTag } from '@local/types';
import { includes, isEmpty, last, pull } from 'lodash';
import React, { useState } from 'react';
import styles from './styles.css';
import Tag from './Tag';
import Tags from './Tags';

interface Props {
  tags: ITag[];
  className?: string;
  onAddTag?: (tag: UserCreatedTag) => void;
  onRemoveTags?: (tags: ITag[]) => void;
}

function TagsEditor({ tags, onAddTag, onRemoveTags, className }: Props) {
  const [selectedTags, setSelectedTags] = useState<ID[]>([]);

  function handleSelect(isSelected: boolean, tag: ITag) {
    if (isSelected) {
      setSelectedTags([...selectedTags, tag.id]);
    } else {
      setSelectedTags(pull(selectedTags, tag.id));
    }
  }

  function handleChange(e: React.ChangeEvent<HTMLInputElement>) {
    if (onAddTag) {
      const tagName = e.target.value;
      if (isValidTagName(tagName)) {
        onAddTag({ name: normalizeTagName(tagName) });
        e.target.value = '';
      }
    }
  }

  function handleRemove(e: React.KeyboardEvent<HTMLInputElement>) {
    if (onRemoveTags && !isEmpty(tags) && isInputEmpty(e)) {
      onRemoveTags(matchTagsToSelected(selectedTags, tags));
      setSelectedTags([]);
    }
  }

  function handleKeyPress(e: React.KeyboardEvent<HTMLInputElement>) {
    const key = e.keyCode || e.charCode;
    if (key === Keys.enter) {
      if (e.currentTarget.value.trim() !== '') {
        if (onAddTag) {
          onAddTag({ name: e.currentTarget.value });
          resetInput(e);
          return;
        }
      }
    }
    if (key === Keys.backspace) {
      handleRemove(e);
      return;
    }
  }

  return (
    <div className={className}>
      <Tags
        tags={tags}
        renderTag={({ tag }) => (
          <Tag
            tag={tag}
            selected={isTagSelected(tag, selectedTags)}
            onSelect={handleSelect}
          />
        )}
      />
      <input
        placeholder="Tag..."
        type="text"
        onChange={handleChange}
        onKeyDown={handleKeyPress}
        className={styles.input}
        data-testid="input"
      />
    </div>
  );
}

function normalizeTagName(name: string) {
  return name.trim().replace(',', '');
}

function isValidTagName(name: string) {
  return name.match(Regex.tagName);
}

function isTagSelected(tag: ITag, selectedTags: ID[]) {
  return includes(selectedTags, tag.id);
}

function matchTagsToSelected(selectedTags: ID[], tags: ITag[]) {
  if (isEmpty(selectedTags)) {
    return [last(tags)!];
  }
  return tags.reduce(
    (matched: ITag[], current: ITag) =>
      includes(selectedTags, current.id) ? [...matched, current] : matched,
    [],
  );
}

function resetInput(e: React.KeyboardEvent<HTMLInputElement>) {
  e.currentTarget.value = '';
}

function isInputEmpty(e: React.KeyboardEvent<HTMLInputElement>) {
  return e.currentTarget.value === '';
}

export default React.memo(TagsEditor);
