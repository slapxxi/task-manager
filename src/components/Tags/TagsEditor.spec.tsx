import { Keys } from '@lib';
import React from 'react';
import { fireEvent, render } from 'react-testing-library';
import TagsEditor from './TagsEditor';

const tags = [{ id: 'html', name: 'HTML' }, { id: 'css', name: 'CSS' }];

it('renders', () => {
  const { container } = render(<TagsEditor tags={tags} />);
  expect(container.firstChild).toMatchSnapshot();
});

it('invokes callback when tag removed', () => {
  const spy = jest.fn();
  const { getByTestId } = render(<TagsEditor tags={tags} onRemoveTag={spy} />);
  fireEvent.keyDown(getByTestId('input'), { keyCode: Keys.backspace });
  expect(spy).toHaveBeenCalledTimes(1);
});

it('removes last tag when no tags selected', () => {
  const spy = jest.fn();
  const { getByTestId } = render(<TagsEditor tags={tags} onRemoveTag={spy} />);
  fireEvent.keyDown(getByTestId('input'), { keyCode: Keys.backspace });
  expect(spy).toHaveBeenCalledWith({ id: 'css', name: 'CSS' });
});

it('removes selected tag', () => {
  const spy = jest.fn();
  const { getByText, getByTestId } = render(
    <TagsEditor tags={tags} onRemoveTag={spy} />,
  );
  fireEvent.click(getByText('HTML'));
  fireEvent.keyDown(getByTestId('input'), { keyCode: Keys.backspace });
  expect(spy).toHaveBeenCalledWith({ id: 'html', name: 'HTML' });
});

it('removes multiple selected tags', () => {
  const spy = jest.fn();
  const { getByText, getByTestId } = render(
    <TagsEditor tags={tags} onRemoveTags={spy} />,
  );
  fireEvent.click(getByText('HTML'));
  fireEvent.click(getByText('CSS'));
  fireEvent.keyDown(getByTestId('input'), { keyCode: Keys.backspace });
  expect(spy).toHaveBeenCalledTimes(1);
  expect(spy).toHaveBeenCalledWith([
    { id: 'html', name: 'HTML' },
    { id: 'css', name: 'CSS' },
  ]);
});

it('does not invoke both callbacks', () => {
  const singleRemoveSpy = jest.fn();
  const multipleRemoveSpy = jest.fn();
  const { getByText, getByTestId } = render(
    <TagsEditor
      tags={tags}
      onRemoveTag={singleRemoveSpy}
      onRemoveTags={multipleRemoveSpy}
    />,
  );
  fireEvent.click(getByText('HTML'));
  fireEvent.click(getByText('CSS'));
  fireEvent.keyDown(getByTestId('input'), { keyCode: Keys.backspace });
  expect(singleRemoveSpy).not.toHaveBeenCalled();
  expect(multipleRemoveSpy).toHaveBeenCalledTimes(1);
});
