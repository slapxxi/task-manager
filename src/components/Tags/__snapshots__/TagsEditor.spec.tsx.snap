// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div>
  <ul
    class="tags"
    data-testid="tags"
  >
    <li
      class="tag"
    >
      <input
        class="checkbox"
        data-testid="checkbox"
        id="-html"
        type="checkbox"
      />
      <label
        class="label"
        for="-html"
      >
        HTML
      </label>
    </li>
    <li
      class="tag"
    >
      <input
        class="checkbox"
        data-testid="checkbox"
        id="-css"
        type="checkbox"
      />
      <label
        class="label"
        for="-css"
      >
        CSS
      </label>
    </li>
  </ul>
  <input
    class="input"
    data-testid="input"
    placeholder="Tag..."
    type="text"
  />
</div>
`;
